RenderStatePackager/src/RenderStatePackager.cpp

```diff
@@ -25,12 +25,156 @@
 
 namespace Diligent
 {
 
+struct BytecodeDumper
+{
+    class WorkingDirectory
+    {
+    public:
+        WorkingDirectory(const Char* Path) :
+            m_CurrentPath{Path} {}
+
+
+        void Push(const Char* Path)
+        {
+            m_CurrentPath = m_CurrentPath + FileSystem::GetSlashSymbol() + Path;
+
+            if (!FileSystem::PathExists(m_CurrentPath.c_str()))
+                FileSystem::CreateDirectory(m_CurrentPath.c_str());
+        }
+
+        void Pop()
+        {
+            m_CurrentPath = m_CurrentPath.substr(0, m_CurrentPath.find_last_of((FileSystem::GetSlashSymbol())));
+        }
+
+        String ComputePathFor(const Char* FileName) const
+        {
+            return m_CurrentPath + FileSystem::GetSlashSymbol() + FileName;
+        }
+
+    private:
+        String m_CurrentPath;
+    };
+
+    class WorkingDirectoryScope
+    {
+    public:
+        WorkingDirectoryScope(WorkingDirectory& Directory, const Char* Path) :
+            m_Directory{Directory}
+        {
+            m_Directory.Push(Path);
+        }
+
+        ~WorkingDirectoryScope()
+        {
+            m_Directory.Pop();
+        }
+
+    private:
+        WorkingDirectory& m_Directory;
+    };
+
+
+    static bool Execute(const std::vector<RefCntAutoPtr<IPipelineState>>& Pipelines, ARCHIVE_DEVICE_DATA_FLAGS DeviceFlags, const char* Path)
+    {
+        auto GetFileExtension = [](ARCHIVE_DEVICE_DATA_FLAGS DeviceFlag, SHADER_SOURCE_LANGUAGE Language, bool UseBytecode) //
+        {
+            if (UseBytecode)
+            {
+                switch (DeviceFlag)
+                {
+                    case ARCHIVE_DEVICE_DATA_FLAG_D3D11:
+                    case ARCHIVE_DEVICE_DATA_FLAG_D3D12:
+                        return ".dxbc";
+                    case ARCHIVE_DEVICE_DATA_FLAG_VULKAN:
+                        return ".spv";
+                    case ARCHIVE_DEVICE_DATA_FLAG_METAL_IOS:
+                    case ARCHIVE_DEVICE_DATA_FLAG_METAL_MACOS:
+                        return ".air";
+                    default:
+                        UNEXPECTED("Unexpected device data flag (", static_cast<Uint32>(DeviceFlag), ")");
+                        return "";
+                }
+            }
+            else
+            {
+                switch (Language)
+                {
+                    case SHADER_SOURCE_LANGUAGE_HLSL:
+                        return ".hlsl";
+                    case SHADER_SOURCE_LANGUAGE_GLSL:
+                    case SHADER_SOURCE_LANGUAGE_GLSL_VERBATIM:
+                        return ".glsl";
+                    case SHADER_SOURCE_LANGUAGE_MSL:
+                        return ".msl";
+                    default:
+                        UNEXPECTED("Unexpected source language (", static_cast<Uint32>(Language), ")");
+                        return "";
+                }
+            }
+        };
+
+        try
+        {
+            WorkingDirectory RootDirectory{Path};
```

Comment:
What happens if path ends with slash?
What happens if the path ends with two slashes?
What happens if the path ends with the '/' on Windows?